### ISSUE TITLE
esp8266: development list

### ISSUE THREAD

**Issue by alex-march** [alex-march/micropython#1]
- [ESP HAL] (esp8266/esp_mphal.c)
- [ ] I2C
- [ ] PWM
- [FATFS] (esp8266/fatfs/src)
- [ ] Finish the port

**Comment by pfalcon** [comment]
FYI, I'm going to implement generic bitbanging I2C. For *MicroPython*, in full accordance with my comment https://github.com/micropython/micropython/pull/1507#issuecomment-152755885 , so it will benefit any port, including esp8266. Well, it will benefit it if it'll have new hardware API ("machine" module), and that's where others can help.

**Comment by alex-march** [comment]
Okay, that's good to know. I will keep an eye out for it.

**Comment by alex-march** [comment]
A few notes on using fsusermount from stmhal in ESP port. I can compile and add the pyb.mount to the existing modpyb as you suggested. This is more for myself, but if you're here and have time for a discussion, would be great.

What can be reused, and can be deleted right now from esp8266/fatfs:
- ffconf.c / ffconf.h
- diskio.h

Conflicts:
 - [stmhal diskio.c#L33](https://github.com/micropython/micropython/blob/master/stmhal/diskio.c#L33): conflicts by including STM32_HAL_H but otherwise can be adapted for esp8266 port, although requires renaming existing sflash_ functions to match and . Can this be replaced with MICROPY_HAL_H since mphal.h includes the STM32_HAL_H?

Refactor:
- sflash_diskio.c/h: to follow the names in stmhal/storage.h and renamed appropriately
- diskio.c: to use fs_user_mount if not possible to reuse stmhal version
- dummy sdcard

All this will leave just 2 files that interface with ESP SDK spi functions and provide the buffer. After some thought, I think you're right about using this approach and the changes seem fairly trivial.

**Comment by pfalcon** [comment]
> After some thought, I think you're right about using this approach and the changes seem fairly trivial.

I'm glad if you find it such.

> This probably won't benefit micropython as such, but will reduce the footprint of my PR to only two files.

Well, it will definitely benefit by testing pyb.mount support for real-world usage and fixing any non-portability in its implementation.

Unfortunately, I cannot comment on specifics above - I have too many open threads of work, I cannot start looking closely into something else. If you find specific issue, please open a ticket in the main tracker.

But generally, there will be 2 independent tasks:

1. Provide Python-level functions to access esp8266 flash. Just as I suggested to do initially, but as you already implemented all that page caching, perhaps working not in terms of sectors, but in terms of byte addressing. So, there should be functions to read X bytes at flash offset Y. Ditto for writing.
2. pyb.mount implementation should be factored out to be reusable across port (and the first target for reuse should be not even esp8266, but unix port, as a common ground and easy testbed).

If you decide to follow this plan, I'll be happy to help as time permits (the key to quick turnaround is small changes).

**Comment by alex-march** [comment]
Hi @pfalcon. Now that the fatfs looks a bit cleaner I am moving on with the pyb.mount dev. I have a question regarding the points you made in the last post, since I'm still new to the project, I thought it's best to consult you before potentially wasting my effort:

1. Should the flash access functions be exposed as a separate module, say modsflash.c which in turn can be used in user mounting? I want to bridge the gap between what I have so far and mounting, so I can start on byte-by-byte operations.

2. To have the mount implementation reusable, would it be appropriate to name the driver class simply `mount` or `blockdev` or something along those lines, what's your opinion?

**Comment by pfalcon** [comment]
> Now that the fatfs looks a bit cleaner I am moving on with the pyb.mount dev.

Yes, thanks for that work!

> Should the flash access functions be exposed as a separate module

Nope, "esp" module is catch-all for esp8266-specific functionality, so just put it there, naming `read_flash()`/`write_flash()` or something.

> To have the mount implementation reusable, would it be appropriate to name the driver class simply mount or blockdev or something along those lines, what's your opinion? 

I'm not sure I understand this question. If you mean a class whose instance will be passed to pyb.mount(), then it's not that important how to name at this point - it's going to be in python, so will be easy to change later.

Please mind the suggestion to expose pyb.mount functionality in unix port. That will require similar legwork as to generalize pyb.mount like you already did, but again will make whole uPy better and will allow to let you play with pyb.mount functionality much easier a get good feel of it, so hopefully everything will "just work" on esp8266 later.

**Comment by alex-march** [comment]
> Nope, "esp" module is catch-all for esp8266-specific functionality, so just put it there, naming read_flash()/write_flash() or something.

Okay, got it!
> class whose instance will be passed to pyb.mount()

Yes, that's what I meant, but you're right, it is not essential right now. I will just crack on with the esp flash operations.

I think following with the unix port first will be a solid option, as I don't have to be the only one testing the changes with my esp, despite all the legwork it should prove worthwhile. Thanks for the tips!

**Comment by pfalcon** [comment]
Glad that you find this plan worthwhile. Btw, did you build the port against 1.4.0 vendor sdk (as available in the latest esp-open-sdk)? I still didn't test it against uPy ;-(

**Comment by alex-march** [comment]
Yeah, I updated to 1.4.0 the other day, no major issues so far, I didn't check the espressiff changelog to be honest. Oh and the esptool.py got updated the other day, the cryptic issue when combining the elf file, see [upstream] (https://github.com/themadinventor/esptool/issues/66). So might be worth updating the submodule in your esp-open-sdk :)

**Comment by pfalcon** [comment]
Thanks for confirming. All that is on to-do, but I'm pretty busy these weeks, so it may be in the backlog for a month or so.

**Comment by alex-march** [comment]
Can't blame you, I've got my full-time job to keep up with as well! I'll keep you posted on the progress.

**Comment by alex-march** [comment]
Hey @pfalcon.
I have the following:
- Flash type inside the esp with functions:
```python
from esp import flash
flash.init() # -> bool
flash.status() # -> bool
flash.read(addr, size) # -> vstr byte array
flash.write(addr, buff) # -> const_none, should be boolean maybe?
flash.sync() # -> bool
```
If I don't use a separate flash type, and just add the functions to esp, such as esp.flash_read(addr, size), everything works fine. However, with the example above, I get `NameError: name not defined
`. Identical functions and parameters, just inside another type. If you have a moment, could you please glance at df9259f. Aside from that, this is probably ready to go into the new PR.

**Comment by pfalcon** [comment]
>  just inside another type

Please don't. There's no need for such complications (unless proved otherwise). So, it should be:

~~~~
import esp
flash_init() # -> bool
~~~~

**Comment by alex-march** [comment]
Okay, I'll move them up into the module then. Since there are a few
functions, I thought it makes sense instead of naming all of them flash_.
Thanks for the advice, I'll do that asap.
